x-pack/plugins/security_solution/public/resolver/view/resolver_without_providers.tsx

```diff
@@ -0,0 +1,136 @@
+/*
```

Comment:
formerly `map.tsx`. i changed a few things but its mostly whitespace. github isn't behaving, but git can:
via `git diff -w upstream/master:./x-pack/plugins/security_solution/public/resolver/view/map.tsx ./x-pack/plugins/security_solution/public/resolver/view/resolver_without_providers.tsx`

```diff
diff --git a/x-pack/plugins/security_solution/public/resolver/view/map.tsx b/x-pack/plugins/security_solution/public/resolver/view/resolver_without_providers.tsx
index 30aa4b63a13..8813e09932a 100644
--- a/x-pack/plugins/security_solution/public/resolver/view/map.tsx
+++ b/x-pack/plugins/security_solution/public/resolver/view/resolver_without_providers.tsx
@@ -8,7 +8,7 @@
 
 /* eslint-disable react/display-name */
 
-import React, { useContext } from 'react';
+import React, { useContext, useCallback } from 'react';
 import { useSelector } from 'react-redux';
 import { useEffectOnce } from 'react-use';
 import { EuiLoadingSpinner } from '@elastic/eui';
@@ -24,30 +24,16 @@ import { useResolverQueryParams } from './use_resolver_query_params';
 import { StyledMapContainer, StyledPanel, GraphContainer } from './styles';
 import { entityId } from '../../../common/endpoint/models/event';
 import { SideEffectContext } from './side_effect_context';
+import { ResolverProps } from '../types';
 
 /**
  * The highest level connected Resolver component. Needs a `Provider` in its ancestry to work.
  */
-export const ResolverMap = React.memo(function ({
-  className,
-  databaseDocumentID,
-  resolverComponentInstanceID,
-}: {
-  /**
-   * Used by `styled-components`.
-   */
-  className?: string;
-  /**
-   * The `_id` value of an event in ES.
-   * Used as the origin of the Resolver graph.
-   */
-  databaseDocumentID?: string;
-  /**
-   * A string literal describing where in the app resolver is located,
-   * used to prevent collisions in things like query params
-   */
-  resolverComponentInstanceID: string;
-}) {
+export const ResolverWithoutProviders = React.memo(
+  React.forwardRef(function (
+    { className, databaseDocumentID, resolverComponentInstanceID }: ResolverProps,
+    refToForward
+  ) {
     /**
      * This is responsible for dispatching actions that include any external data.
      * `databaseDocumentID`
@@ -63,12 +49,28 @@ export const ResolverMap = React.memo(function ({
       selectors.visibleNodesAndEdgeLines
     )(timeAtRender);
     const terminatedProcesses = useSelector(selectors.terminatedProcesses);
-  const { projectionMatrix, ref, onMouseDown } = useCamera();
+    const { projectionMatrix, ref: cameraRef, onMouseDown } = useCamera();
+
+    const ref = useCallback(
+      (element: HTMLDivElement | null) => {
+        // Supply `useCamera` with the ref
+        cameraRef(element);
+
+        // If a ref is being forwarded, populate that as well.
+        if (typeof refToForward === 'function') {
+          refToForward(element);
+        } else if (refToForward !== null) {
+          refToForward.current = element;
+        }
+      },
+      [cameraRef, refToForward]
+    );
     const isLoading = useSelector(selectors.isLoading);
     const hasError = useSelector(selectors.hasError);
     const activeDescendantId = useSelector(selectors.ariaActiveDescendant);
     const { colorMap } = useResolverTheme();
     const { cleanUpQueryParams } = useResolverQueryParams();
+
     useEffectOnce(() => {
       return () => cleanUpQueryParams();
     });
@@ -76,11 +78,11 @@ export const ResolverMap = React.memo(function ({
     return (
       <StyledMapContainer className={className} backgroundColor={colorMap.resolverBackground}>
         {isLoading ? (
-        <div className="loading-container">
+          <div data-test-subj="resolver:graph:loading" className="loading-container">
             <EuiLoadingSpinner size="xl" />
           </div>
         ) : hasError ? (
-        <div className="loading-container">
+          <div data-test-subj="resolver:graph:error" className="loading-container">
             <div>
               {' '}
               <FormattedMessage
@@ -91,6 +93,7 @@ export const ResolverMap = React.memo(function ({
           </div>
         ) : (
           <GraphContainer
+            data-test-subj="resolver:graph"
             className="resolver-graph kbn-resetFocusState"
             onMouseDown={onMouseDown}
             ref={ref}
@@ -98,7 +101,8 @@ export const ResolverMap = React.memo(function ({
             tabIndex={0}
             aria-activedescendant={activeDescendantId || undefined}
           >
-          {connectingEdgeLineSegments.map(({ points: [startPosition, endPosition], metadata }) => (
+            {connectingEdgeLineSegments.map(
+              ({ points: [startPosition, endPosition], metadata }) => (
                 <EdgeLine
                   edgeLineMetadata={metadata}
                   key={metadata.uniqueId}
@@ -106,7 +110,8 @@ export const ResolverMap = React.memo(function ({
                   endPosition={endPosition}
                   projectionMatrix={projectionMatrix}
                 />
-          ))}
+              )
+            )}
             {[...processNodePositions].map(([processEvent, position]) => {
               const processEntityId = entityId(processEvent);
               return (
@@ -127,4 +132,5 @@ export const ResolverMap = React.memo(function ({
         <SymbolDefinitions />
       </StyledMapContainer>
     );
-});
+  })
+);
```

Reply:
Nice, I would add a comment here that the ref callback is used in the tests etc..., even though it has more use cases than that.

Reply:
did